File: python/server-maintenance.py
#!/usr/bin/env python3
# server-maintenance.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway1:scripts/server-maintenance.sh
# auch in: OpenClaw@gateway2:scripts/server-maintenance.sh
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

import subprocess
import datetime
import os
import sys

# Server Maintenance Script
# RAM: 8GB, Uhr: Europe/Berlin

LOG_FILE = "/var/log/server-maintenance.log"
DATE = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
HOST = os.uname()[1]

# Farben für Terminal
RED = '\033[0;31m'
GREEN = '\033[0;32m'
YELLOW = '\033[1;33m'
NC = '\033[0m'

def log_message(message):
    """Log message to file and print to console"""
    formatted_message = f"[{DATE}] {message}"
    print(formatted_message)
    with open(LOG_FILE, 'a') as f:
        f.write(formatted_message + '\n')

log_message("=== Server Maintenance Check ===")

# 1. APT Update Check
log_message("Checking for updates...")
try:
    result = subprocess.run(['apt', 'update', '-qq'], capture_output=True, text=True)
    lines = result.stdout.split('\n')[-6:-1]
    for line in lines:
        if line.strip():
            log_message(line.strip())
    
    result = subprocess.run(['apt', 'list', '--upgradable'], capture_output=True, text=True)
    updates = len(result.stdout.split('\n')) - 1
    if updates > 1:
        log_message(f"⚠️ {updates} packages can be upgraded")
except Exception as e:
    log_message(f"Error checking updates: {e}")

# 2. RAM Check (8GB total)
log_message("Checking RAM usage...")
RAM_TOTAL = 8192  # 8GB in MB
try:
    result = subprocess.run(['free', '-m'], capture_output=True, text=True)
    lines = result.stdout.split('\n')
    mem_line = [line for line in lines if line.startswith('Mem:')][0]
    parts = mem_line.split()
    RAM_USED = int(parts[2])
    RAM_PERCENT = (RAM_USED * 100) // RAM_TOTAL
    log_message(f"RAM: {RAM_USED}MB / {RAM_TOTAL}MB ({RAM_PERCENT}%)")
    if RAM_PERCENT > 90:
        log_message("🔴 WARNING: RAM usage > 90%!")
    elif RAM_PERCENT > 80:
        log_message("🟡 WARNING: RAM usage > 80%")
except Exception as e:
    log_message(f"Error checking RAM: {e}")

# 3. Disk Space Check
log_message("Checking disk space...")
try:
    result = subprocess.run(['df', '-h', '/'], capture_output=True, text=True)
    df_line = result.stdout.split('\n')[1]
    parts = df_line.split()
    disk_info = f"Disk: {parts[2]} / {parts[1]} ({parts[4]} used)"
    log_message(disk_info)
    
    result = subprocess.run(['df', '/'], capture_output=True, text=True)
    df_line = result.stdout.split('\n')[1]
    parts = df_line.split()
    DISK_PERCENT = int(parts[4].rstrip('%'))
    if DISK_PERCENT > 90:
        log_message("🔴 WARNING: Disk > 90%!")
    elif DISK_PERCENT > 80:
        log_message("🟡 WARNING: Disk > 80%")
except Exception as e:
    log_message(f"Error checking disk space: {e}")

# 4. NTP Check
log_message("Checking NTP sync...")
try:
    result = subprocess.run(['timedatectl', 'status'], capture_output=True, text=True)
    if "NTP synchronized: yes" in result.stdout:
        log_message("✅ NTP synchronized")
    else:
        log_message("⚠️ NTP not synchronized")
except Exception as e:
    log_message(f"Error checking NTP: {e}")

# 5. OpenClaw Gateway Status
log_message("Checking OpenClaw Gateway...")
try:
    result = subprocess.run(['systemctl', 'is-active', 'openclaw-gateway'], 
                          capture_output=True, text=True)
    if result.returncode == 0 and result.stdout.strip() == 'active':
        log_message("✅ OpenClaw Gateway running")
    else:
        log_message("🔴 OpenClaw Gateway NOT running!")
        subprocess.run(['systemctl', 'restart', 'openclaw-gateway'])
except Exception as e:
    log_message(f"Error checking OpenClaw Gateway: {e}")

# 6. Load Average
try:
    result = subprocess.run(['uptime'], capture_output=True, text=True)
    load_part = result.stdout.split('load average:')[1].strip()
    LOAD = load_part.split(',')[0].strip()
    log_message(f"Load Average: {LOAD}")
except Exception as e:
    log_message(f"Error checking load average: {e}")

log_message("=== Maintenance Complete ===")
log_message("")

File: javascript/server-maintenance.js
#!/usr/bin/env node
// server-maintenance.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway1:scripts/server-maintenance.sh
// auch in: OpenClaw@gateway2:scripts/server-maintenance.sh
// Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

// Server Maintenance Script
// RAM: 8GB, Uhr: Europe/Berlin

const fs = require('fs');
const { execSync } = require('child_process');

const LOG_FILE = "/var/log/server-maintenance.log";
const DATE = new Date().toISOString().replace('T', ' ').substring(0, 19);
const HOST = execSync('hostname').toString().trim();

// Farben für Terminal
const RED = '\x1b[31m';
const GREEN = '\x1b[32m';
const YELLOW = '\x1b[33m';
const NC = '\x1b[0m';

function log(message) {
    const logMessage = `[${DATE}] ${message}`;
    console.log(logMessage);
    fs.appendFileSync(LOG_FILE, logMessage + '\n');
}

log("=== Server Maintenance Check ===");

// 1. APT Update Check
log("Checking for updates...");
try {
    const aptUpdateOutput = execSync('apt update -qq 2>&1', { encoding: 'utf8' });
    const lastLines = aptUpdateOutput.split('\n').slice(-5).join('\n');
    log(lastLines);
    
    const upgradablePackages = execSync('apt list --upgradable 2>/dev/null', { encoding: 'utf8' });
    const updates = upgradablePackages.trim() ? upgradablePackages.split('\n').length - 1 : 0;
    
    if (updates > 0) {
        log(`⚠️ ${updates} packages can be upgraded`);
    }
} catch (error) {
    log(`Error checking updates: ${error.message}`);
}

// 2. RAM Check (8GB total)
log("Checking RAM usage...");
const RAM_TOTAL = 8192; // 8GB in MB
try {
    const freeOutput = execSync('free -m', { encoding: 'utf8' });
    const ramLine = freeOutput.split('\n')[1];
    const ramUsed = parseInt(ramLine.split(/\s+/)[2]);
    const ramPercent = Math.round((ramUsed * 100) / RAM_TOTAL);
    
    log(`RAM: ${ramUsed}MB / ${RAM_TOTAL}MB (${ramPercent}%)`);
    
    if (ramPercent > 90) {
        log("🔴 WARNING: RAM usage > 90%!");
    } else if (ramPercent > 80) {
        log("🟡 WARNING: RAM usage > 80%");
    }
} catch (error) {
    log(`Error checking RAM: ${error.message}`);
}

// 3. Disk Space Check
log("Checking disk space...");
try {
    const dfOutput = execSync('df -h /', { encoding: 'utf8' });
    const diskInfo = dfOutput.split('\n')[1].split(/\s+/);
    const used = diskInfo[2];
    const total = diskInfo[1];
    const percent = diskInfo[4];
    
    log(`Disk: ${used} / ${total} (${percent} used)`);
    
    const diskPercent = parseInt(percent.replace('%', ''));
    if (diskPercent > 90) {
        log("🔴 WARNING: Disk > 90%!");
    } else if (diskPercent > 80) {
        log("🟡 WARNING: Disk > 80%");
    }
} catch (error) {
    log(`Error checking disk space: ${error.message}`);
}

// 4. NTP Check
log("Checking NTP sync...");
try {
    const ntpStatus = execSync('timedatectl status', { encoding: 'utf8' });
    if (ntpStatus.includes("NTP synchronized: yes")) {
        log("✅ NTP synchronized");
    } else {
        log("⚠️ NTP not synchronized");
    }
} catch (error) {
    log(`Error checking NTP: ${error.message}`);
}

// 5. OpenClaw Gateway Status
log("Checking OpenClaw Gateway...");
try {
    execSync('systemctl is-active --quiet openclaw-gateway');
    log("✅ OpenClaw Gateway running");
} catch (error) {
    log("🔴 OpenClaw Gateway NOT running!");
    try {
        execSync('systemctl restart openclaw-gateway');
    } catch (restartError) {
        log(`Failed to restart OpenClaw Gateway: ${restartError.message}`);
    }
}

// 6. Load Average
try {
    const uptimeOutput = execSync('uptime', { encoding: 'utf8' });
    const loadAvg = uptimeOutput.split('load average:')[1].trim().split(',')[0].trim();
    log(`Load Average: ${loadAvg}`);
} catch (error) {
    log(`Error checking load average: ${error.message}`);
}

log("=== Maintenance Complete ===");
log("");
